### ISSUE TITLE
Running AESM without internet connectivity

### ISSUE THREAD

**Issue by rdockterjr** [intel/linux-sgx#799]
Currently I am running AESM and an SGX application in docker containers similar to the examples here: https://github.com/intel/linux-sgx/tree/master/docker/build. I should note I am using hardware release mode and have a valid key in the launch policy whitelist.

The docker containers come up just fine when the system has an internet connection. However in scenarios where this machine does not have internet access, the AESM service is not able to update the whitelist from whitelist.trustedservices.intel.com
I get the following error after some time:

> aesm-service          | aesm_service[7]: The server sock is 0x563e3a15a640
> aesm-service          | aesm_service[7]: [ADMIN]White List update requested
> aesm-service          | aesm_service[7]: [ADMIN]White List update failed due to network error

I don't require internet access otherwise since I am just using local attestation. I am wondering if there is any way to use a cached version of the the sgx_white_list_cert.bin file instead of always updating when the AESM container starts up.

I noticed in the aesmd.conf file there is a "whitelist url" variable that can be set. Would it work to host the bin file locally and point the url there? I couldn't find much documentation for the conf file.

Any suggestions would be much appreciated, or is it just not feasible to run SGX completely offline?

**Comment by llly** [comment]
It's OK if white list is not updated. There is already a cached version white list when installing AESM. 
If you want manually download white list and provide it to AESM, 
 download http://whitelist.trustedservices.intel.com/SGX/LCWL/Linux/sgx_white_list_cert.bin, replace 
`/var/opt/aesmd/data/white_list_cert_to_be_verify.bin` as root, then restart AESM. You will still see `update failed` error.
This URL locates at https://github.com/intel/linux-sgx/blob/master/psw/ae/data/constants/linux/default_url_info.hh

**Comment by rdockterjr** [comment]
Thanks for the response.

Based on your approach it sounds like I could achieve the same thing by downloading sgx_white_list_cert.bin and mounting it into the aesm docker container like so:

```
aesm-service:
  image: aesm-service
  devices:
    - /dev/isgx
  volumes:
    - sgx_white_list_cert.bin:/var/opt/aesmd/data/white_list_cert_to_be_verify.bin
    - aesmd-socket:/var/run/aesmd
```

I will run some tests now.

**Comment by rdockterjr** [comment]
I was able to confirm that mounting sgx_white_list_cert.bin into the aesm docker container allows the aesm service to load up properly without internet connectivity.
Without this file though the aesm service does not run properly and my sgx application cannot load the enclave. 
My guess is that inside docker the aesm service is unable to manually cache a version of the whitelist. Therefore it must be provided one.
Closing the issue.
